infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/ILogReplicationVersionAdapter.java

```diff
@@ -12,4 +14,13 @@ public interface ILogReplicationVersionAdapter {
      * @return Version string that indicates the version of LR.
      */
     String getVersion();
+
+    /**
+     * @param txnContext - transaction context where we are checking if we need
+     *                   to behave in a backward compatible manner.
+     * @return true if the cluster is undergoing rolling upgrades with
+     * some nodes running an older code base.
+     * false if all the nodes in the cluster are at the same version as me.
+     */
+    boolean isRollingUpgradeInProgress(TxnContext txnContext);
```

Comment:
I think that the method to be exposed here should be getClusterVersion().

isRollingUpgradeInProgress() is our internal method which checks the following:
1) all nodes in the cluster are on the same version?  (depends on Chris's change to add version info per node)
2) node version == clusterVersion (returned from the plugin).

All places in LR invoke isRollingUpgradeInProgress() to decide whether to use the old or new api's.

Additionally, on startup, LR starts a periodic thread which calls isRollingUpgradeInProgress().  If not, and if the atomic boolean is not set, it clears all old data and updates the atomic boolean.  Also, stops this executor service as migration/clear of old data is already done.

Reply:
Done, please checkout latest commit. thanks